### PR TITLE
Add support for MongoDB Atlas $vectorSearch vector search

### PR DESCRIPTION
**Pull request by NoahStapp**
# Description

Adds support for the $vectorSearch operator for MongoDBAtlasVectorSearch, which was announced at .Local London (September 26th, 2023). This change breaks backwards compatibility support for the existing $search operator used by the original integration (https://github.com/jerryjliu/llama_index/pull/6379).

## Type of Change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Added new unit/integration tests

# Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: llama_index/vector_stores/mongodb.py [hunk 1]
```diff
@@ -33,19 +35,12 @@ def _to_mongodb_filter(standard_filters: MetadataFilters) -> Dict:
...
-
-    Args:
-        mongodb_index (Optional[pymongo.MongoClient]): MongoDB index instance
-        db_name (str): MongoDB database name
-        collection_name (str): MongoDB collection name
-        insert_kwargs (Optional[Dict]): kwargs used during `insert`.
+    To use, you should have both:
+    - the ``pymongo`` python package installed
+    - a connection string associated with a MongoDB Atlas Cluster
+    that has a Atlas Vector Search index
```
Comment (blink1073) [inline llama_index/vector_stores/mongodb.py:43]:
```suggestion
    that has an Atlas Vector Search index
```

File: llama_index/vector_stores/mongodb.py [hunk 2]
```diff
@@ -65,7 +60,17 @@ def __init__(
         insert_kwargs: Optional[Dict] = None,
         **kwargs: Any,
     ) -> None:
-        """Initialize params."""
+        """Args:
+        mongodb_client: A MongoDB client.
```
Comment (blink1073) [inline llama_index/vector_stores/mongodb.py:64]: I'd suggest "Initialize the vector store", and then format the `Args:` block consistent with the other functions.

File: llama_index/vector_stores/mongodb.py [hunk 3]
```diff
@@ -203,3 +228,30 @@ def query(self, query: VectorStoreQuery, **kwargs: Any) -> VectorStoreQueryResul
         )
         logger.debug("Result of query: %s", result)
         return result
+
+    def query(self, query: VectorStoreQuery, **kwargs: Any) -> VectorStoreQueryResult:
+        """Query index for top k most similar nodes.
+
+        Args:
+            query: a VectorStoreQuery object
```
Comment (blink1073) [inline llama_index/vector_stores/mongodb.py:236]:
```suggestion
            query: a VectorStoreQuery object.
```

File: llama_index/vector_stores/mongodb.py [hunk 4]
```diff
@@ -203,3 +228,30 @@ def query(self, query: VectorStoreQuery, **kwargs: Any) -> VectorStoreQueryResul
...
+        Returns:
+            A VectorStoreQueryResult containing the results of the query.
+        """
+        if self._use_vectorsearch:
+            try:
+                result = self._query_vectorsearch(query)
+            except OperationFailure as e:
+                # Check for unsupported operation errors
+                if "$vectorSearch" in str(e):
+                    logger.debug(
```
Comment (blink1073) [inline llama_index/vector_stores/mongodb.py:247]: Should this be a higher log level?

**Comment by logan-markewich** [comment]
@blink1073 @NoahStapp Thanks for being on the ball with adding this! I googled it very quicky, but can you summarize what the differences/improvements are here?

Previously, I know you had to manually create a vector search index in the GUI after uploading/inserting your data before you could query. Is this still the case with these updates?

File: llama_index/vector_stores/mongodb.py (hunk 4 above)
Comment (NoahStapp) [reply llama_index/vector_stores/mongodb.py:247]: Yes, supposed to be `error`. Fixed, thanks!

**Review by blink1073 — APPROVED** [review]
LGTM!

**Comment by logan-markewich** [comment]
@NoahStapp can you fix the unit testing and linting errors?

Linting is fairly easy to fix. Just some lines are too long
```
llama_index/vector_stores/mongodb.py:71:89: E501 Line too long (93 > 88 characters)
llama_index/vector_stores/mongodb.py:73:89: E501 Line too long (91 > 88 characters)
```

For the unit tests, it looks like a typing issue

**Comment by NoahStapp** [comment]
> @NoahStapp can you fix the unit testing and linting errors?
> 
> Linting is fairly easy to fix. Just some lines are too long
> 
> ```
> llama_index/vector_stores/mongodb.py:71:89: E501 Line too long (93 > 88 characters)
> llama_index/vector_stores/mongodb.py:73:89: E501 Line too long (91 > 88 characters)
> ```
> 
> For the unit tests, it looks like a typing issue

Sure, done. I also removed fallback compatibility support for the older implementation due to incompatibility with the Atlas search backend.

**Comment by logan-markewich** [comment]
@NoahStapp is there any migration for users who had existing atlas indexes?

**Comment by NoahStapp** [comment]
> @NoahStapp is there any migration details I should know about for users who had existing atlas indexes?

Existing indexes should work without issue. Any pre-filters users have written for their vector search queries will have to be re-written in MQL, but otherwise existing workflows should be largely compatible.

**Comment by prakul** [comment]
@logan-markewich we would like to having this ready with all the review, but hold off on getting this merged, as we tee off some documentation. Is there a tag we can add (or notify someone) when ready to merge?

**Comment by logan-markewich** [comment]
@prakul yea just ping me when it's ready (either on here or discord 👍🏻 )

**Comment by prakul** [comment]
@logan-markewich We are ready to merge!

**Comment by logan-markewich** [comment]
@prakul can you fix the tests? It looks like you might need a skip decorator, or mock some api calls

**Comment by NoahStapp** [comment]
> @prakul can you fix the tests? It looks like you might need a skip decorator, or mock some api calls

My bad, forgot to skip the Atlas-requiring tests.

**Comment by prakul** [comment]
@logan-markewich The unit tests are fixed and we are good to merge now
